Feature/#4/홈페이지 UI

heyn2 opened the pull request:
## 1. 무슨 이유로 코드를 변경했나요?

<br>

## 2. 관련 스크린샷을 첨부해주세요.
![image](https://github.com/user-attachments/assets/fd2f67f0-03b1-43df-a0c3-7b46cb3b2df4)

![image](https://github.com/user-attachments/assets/d8d35adb-0da6-4955-ba14-9e80cd4b6520)
(게시글 모달에서 게시글 작성 후 게시물이 정렬되지 않는 문제 해결 필요)

<br>

## 3. 완료 사항
1. 홈페이지 레이아웃 구성
2. 사이드바에 네비게이션 링크 추가
3. 게시글 작성 버튼 및 모달 기능 구현
4. 사용자 프로필 정보 표시 기능 추가
<br>

## 4. 추가 사항
Xcellent-FE\src\app\(route)\(afterLogin)\home\PostModal.tsx 추가하여 게시글 모달 생성
<br>

heyn2 committed:
layout.tsx

사이드바 네비게이션 레이아웃 및 홈 레이아웃 구조 추가

- 사이드바에 네비게이션 링크 추가 (홈, 탐색하기, 알림, 쪽지, 리스트, 프로필)
- 사이드바에 게시하기 버튼과 프로필 섹션 추가
- `HomeLayout` 컴포넌트에 children prop을 활용한 메인 레이아웃 구조 구현
- 사이드바와 메인 레이아웃 컴포넌트에 반응형 디자인 적용

heyn2 committed:
page.tsx

홈 페이지 레이아웃 구현 및 게시글, 댓글 기능 추가

- 게시글 작성, 좋아요, 리트윗, 댓글 작성 기능 구현
- 이미지 업로드 기능 추가 (최대 4장 제한)
- 반응형 사이드바와 게시글 정렬 추가
- 텍스트 영역 자동 확장 기능 추가
- 댓글 섹션 토글 기능 및 댓글 입력 기능 구현

heyn2 committed:
홈페이지 UI 구현 및 게시물 작성, 상호작용 기능 추가 (댓글, 좋아요, 리트윗 기능 포함)

heyn2 committed:
홈페이지 레이아웃 구현 및 게시글 작성 모달 기능 추가

yezzan9 commented on src/app/(route)/(afterLogin)/home/layout.tsx:
```diff
@@ -73,8 +89,12 @@ const HomeLayout: React.FC<HomeLayoutProps> = ({ children }) => {
               </li>
             </ul>
```
스타일이 동일하게 반복되는 부분은 객체를 매핑하는 식으로 수정하면 좋을 것 같아요~!

yezzan9 replied:
next.js는 a태그 대신 Link 태그를 사용하는 것이 좋습니다!

yezzan9 commented on src/app/(route)/(afterLogin)/home/layout.tsx:
```diff
@@ -97,6 +117,40 @@ const HomeLayout: React.FC<HomeLayoutProps> = ({ children }) => {
       <main className="flex-1 ml-72">
         {children}
       </main>
+      {isModalOpen && (
+        <div className="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center">
+          <div className="bg-white text-black p-6 rounded-lg w-96">
+            <div className="flex justify-between items-center mb-4">
+              <h2 className="text-xl font-bold">게시글 작성</h2>
+              <button onClick={handleCloseModal} className="text-gray-500 hover:text-gray-700">
+                <svg className="w-6 h-6" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
+                  <path strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="M6 18L18 6M6 6l12 12"></path>
+                </svg>
+              </button>
+            </div>
+            <textarea
+              className="w-full h-32 p-2 border border-gray-300 rounded mb-4"
+              placeholder="무슨 일이 일어나고 있나요?"
+              value={postContent}
+              onChange={handlePostContentChange}
+            />
+            <div className="flex justify-end">
+              <button
+                onClick={handleCloseModal}
+                className="bg-gray-500 hover:bg-gray-600 text-white py-2 px-4 rounded mr-2"
+              >
+                닫기
+              </button>
+              <button
+                onClick={handlePostSubmit}
+                className="bg-blue-500 hover:bg-blue-600 text-white py-2 px-4 rounded"
+              >
+                게시하기
+              </button>
+            </div>
+          </div>
+        </div>
+      )}
```
모달의 경우 별도의 컴포넌트로 분리하는 것이 가독성 및 유지보수 측면에서 유리합니다!
상단의 div태그 두개는 하나로 통합할 수 있을 것 같아요

dalzzy commented on src/app/(route)/(afterLogin)/home/layout.tsx:
```diff
@@ -73,8 +89,12 @@ const HomeLayout: React.FC<HomeLayoutProps> = ({ children }) => {
               </li>
             </ul>
           </nav>
```
왼쪽 메뉴바는 프로필이나 채팅에서도 사용되는 공통 컴포넌트라서 컴포넌트를 재사용할 수 있도록 (afterLogin)/_component에  코드를 분리하는게 좋을 것 같습니다!

dalzzy replied:
프로필을 클릭했을 때 /profile 이 아니라 const me = { id: "dahyeon" }; const <Link href={`/${me?.id}`}> 이런식으로 동적 경로를 사용하도록 수정부탁드립니다.. api 연결 전까지 임시적으로 설정해놓으면 될 것 같아요

heyn2 replied:
(afterLogin)\_component\Sidebar.tsx 만들어서 코드 분리 완료 했고,  두 번째 피드백 주신 부분에서 동적 경로 사용하도록 코드상에 반영 하였는데 여전히 프로필을 클릭하면 http://localhost:3000/profile 이런 경로로 나타납니다.. 자세한 내용은 pr로 올려보도록 하겠습니다..!

dalzzy commented on src/app/(route)/(afterLogin)/home/layout.tsx:
commit 단위를 좀 더 세분화해서 올려주시면 좋을 것 같습니다..!

yezzan9 commented on src/app/(route)/(afterLogin)/home/page.tsx:
```diff
@@ -1,114 +1,329 @@
+}
+
 const Home: React.FC = () => {
   const [posts, setPosts] = useState<Post[]>([]);
+  const [newPost, setNewPost] = useState<string>("");
+  const [selectedImages, setSelectedImages] = useState<string[]>([]);
+  const [showComments, setShowComments] = useState<number | null>(null);
+  const [newComment, setNewComment] = useState<string>("");
+  const [isModalOpen, setIsModalOpen] = useState(false);
```
newPost, newComment 상태는 약간 굳이,, 인 것 같습니다
해당 상태는 로컬에서 임시저장하는 것 같은데 x서비스에는 임시저장 상태로 데이터를 저장하는 기능만 존재합니다 !

showComments도 comments로 수정하면 좋을 것 같아요~🤗

yezzan9 commented on src/app/(route)/(afterLogin)/home/page.tsx:
```diff
@@ -1,114 +1,329 @@
+    if (newPost.trim() || selectedImages.length > 0) {
+      const post: Post = {
+        id: posts.length + 1,
+        author: "Myself",
+        authorImage: "/profile-placeholder.png",
+        content: newPost,
+        images: selectedImages,
+        likes: 0,
+        retweets: 0,
+        comments: [],
         isLiked: false,
-      },
-      // 더 많은 게시글 추가 가능
-    ]);
-  }, []);
+      };
+      setPosts((prevPosts) => [...prevPosts, post].sort((a, b) => b.likes - a.likes));
+      setNewPost("");
+      setSelectedImages([]);
+    }
+  };
```
post, like, retweet, comment 부분은 백엔드와 dto 조율이 된 건가요?-?

yezzan9 replied:
post, like, retweet, comment 부분은 백엔드와 dto 조율이 된 건가요?-?

yezzan9 commented on src/app/(route)/(afterLogin)/home/page.tsx:
```diff
@@ -1,114 +1,329 @@
 
   const sortedPosts = [...posts].sort((a, b) => b.likes - a.likes);
 
   return (
```
컴포넌트 분리가 전혀 안 되어있는 것 같은데, 게시글 내용 입력 부분 / 트윗 등의 단위로 컴포넌트를 분리하는 편이 좋을 것 같습니다!

dalzzy commented on src/app/(route)/(afterLogin)/home/layout.tsx:
```diff
@@ -97,6 +117,40 @@ const HomeLayout: React.FC<HomeLayoutProps> = ({ children }) => {
       <main className="flex-1 ml-72">
         {children}
       </main>
+      {isModalOpen && (
+        <div className="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center">
+          <div className="bg-white text-black p-6 rounded-lg w-96">
+            <div className="flex justify-between items-center mb-4">
+              <h2 className="text-xl font-bold">게시글 작성</h2>
+              <button onClick={handleCloseModal} className="text-gray-500 hover:text-gray-700">
+                <svg className="w-6 h-6" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
+                  <path strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="M6 18L18 6M6 6l12 12"></path>
+                </svg>
+              </button>
+            </div>
+            <textarea
+              className="w-full h-32 p-2 border border-gray-300 rounded mb-4"
+              placeholder="무슨 일이 일어나고 있나요?"
+              value={postContent}
+              onChange={handlePostContentChange}
+            />
+            <div className="flex justify-end">
+              <button
+                onClick={handleCloseModal}
+                className="bg-gray-500 hover:bg-gray-600 text-white py-2 px-4 rounded mr-2"
+              >
+                닫기
+              </button>
+              <button
+                onClick={handlePostSubmit}
+                className="bg-blue-500 hover:bg-blue-600 text-white py-2 px-4 rounded"
+              >
+                게시하기
+              </button>
+            </div>
+          </div>
+        </div>
```
인터셉팅 라우트과 패러렐 라우트를 사용해서 모달을 구현하시면 실제 X의 url과 같이 구현이 가능합니다..! 저도 이 방식으로 모달 구현했어요..! 현재는 useState를 사용해서 모달창을 띄워주고 있어서 실제 X와 url이 다르게 표현되고 있는 것 같아요

yezzan9 replied:
모달의 경우 별도의 컴포넌트로 분리하는 것이 가독성 및 유지보수 측면에서 유리합니다!
상단의 div태그 두개는 하나로 통합할 수 있을 것 같아요

yezzan9 commented on src/app/(route)/(afterLogin)/home/layout.tsx:
```diff
@@ -73,8 +89,12 @@ const HomeLayout: React.FC<HomeLayoutProps> = ({ children }) => {
               </li>
             </ul>
```
스타일이 동일하게 반복되는 부분은 객체를 매핑하는 식으로 수정하면 좋을 것 같아요~!

yezzan9 replied:
next.js는 a태그 대신 Link 태그를 사용하는 것이 좋습니다!

yezzan9 commented on src/app/(route)/(afterLogin)/home/page.tsx:
```diff
@@ -1,114 +1,329 @@
+
 const Home: React.FC = () => {
   const [posts, setPosts] = useState<Post[]>([]);
+  const [newPost, setNewPost] = useState<string>("");
+  const [selectedImages, setSelectedImages] = useState<string[]>([]);
+  const [showComments, setShowComments] = useState<number | null>(null);
+  const [newComment, setNewComment] = useState<string>("");
+  const [isModalOpen, setIsModalOpen] = useState(false);
+  const textareaRef = useRef<HTMLTextAreaElement>(null);
```
newPost, newComment 상태는 약간 굳이,, 인 것 같습니다
해당 상태는 로컬에서 임시저장하는 것 같은데 x서비스에는 임시저장 상태로 데이터를 저장하는 기능만 존재합니다 !

showComments도 comments로 수정하면 좋을 것 같아요~🤗

yezzan9 commented on src/app/(route)/(afterLogin)/home/page.tsx:
```diff
@@ -1,114 +1,329 @@
-    
-
-      {/* Main Content */}
+    <div className="flex min-h-screen bg-black text-white">
```
컴포넌트 분리가 전혀 안 되어있는 것 같은데, 게시글 내용 입력 부분 / 트윗 등의 단위로 컴포넌트를 분리하는 편이 좋을 것 같습니다!

yezzan9 reviewed (changes requested):
오류가 있어서 리뷰 다시 남겨드립니다 !

yezzan9 commented:
현재 PR에 첨부되어있는 스크린샷만 해도 UI가 x서비스와 많이 다르게 구현되어있어요😂
실제 x서비스의 UI를 더 꼼꼼하게 확인하신 후 수정 부탁드립니다!

<img width="605" alt="스크린샷 2024-11-05 오후 11 52 40" src="https://github.com/user-attachments/assets/ca6a77f5-4aea-4b5c-951f-2b4feb256cd1">
<img width="605" alt="스크린샷 2024-11-06 오전 12 42 12" src="https://github.com/user-attachments/assets/9bfab6b1-ae39-4eab-bd89-96de480d7eb5">
<img width="627" alt="스크린샷 2024-11-06 오전 12 43 35" src="https://github.com/user-attachments/assets/94af8672-d42d-4829-aa90-be34c14e94fe">

yezzan9 commented:
커밋 컨벤션이 지켜지지 않고 있어요🥲 팀 노션을 참고하여 커밋 컨벤션을 지켜주세요!
커밋 단위도 현재보다 더 잘게 나누는 걸 추천드립니다!

dalzzy commented:
예진님이 첨부해주신 사진에서 "추천" , "팔로우 중" 탭에서 "팔로우 중" 탭만 구현하시면 됩니다..! 이미 알고 계신지 모르겠지만 혹시나 해서,, 알려드립니당..

dalzzy commented on src/app/(route)/(afterLogin)/home/page.tsx:
```diff
@@ -1,114 +1,329 @@
     );
   };
 
   const handleRetweet = (id: number) => {
```
![image](https://github.com/user-attachments/assets/f59f0d14-2ec3-416a-9669-92cee184cba6)
리트윗 버튼을 누를 때마다 내가 리트윗한 게시물이 아래에 뜨는 방식으로 구현이 된 것 같은데 의도하신걸까용?..  실제 X의 home에서는 리트윗한 게시물이 따로 나타나지 않고 리트윗한 다른 사람의 게시물만 표시되는 방식이 사용되고 있습니다!

dalzzy commented on src/app/(route)/(afterLogin)/home/page.tsx:
```diff
@@ -1,114 +1,329 @@
-      setPosts((prevPosts) => [
-        ...prevPosts,
-        { ...postToRetweet, id: prevPosts.length + 1, author: "Myself" },
-      ]);
+    setPosts((prevPosts) => {
+      const postToRetweet = prevPosts.find((post) => post.id === id);
+      if (postToRetweet) {
+        const updatedPosts = prevPosts.map((post) =>
+          post.id === id ? { ...post, retweets: post.retweets + 1 } : post
+        );
+        return [
+          ...updatedPosts,
+          { ...postToRetweet, id: prevPosts.length + 1, author: "Myself" },
+        ].sort((a, b) => b.likes - a.likes);
+      }
+      return prevPosts;
+    });
+  };
```
![image](https://github.com/user-attachments/assets/f59f0d14-2ec3-416a-9669-92cee184cba6)
리트윗 버튼을 누를 때마다 내가 리트윗한 게시물이 아래에 뜨는 방식으로 구현이 된 것 같은데 의도하신걸까용?..  실제 X의 home에서는 리트윗한 게시물이 따로 나타나지 않고 리트윗한 다른 사람의 게시물만 표시되는 방식이 사용되고 있습니다!

heyn2 committed:
왼쪽 sidebar 컴포넌트 생성

heyn2 committed:
게시글 작성 모달

heyn2 committed:
isOpen 속성이 PostModalProps 타입에 정의되어 있지 않아 isOpen을 사용할 수 없다는 오류발생

heyn2 committed:
사이드바 변경, 프로필 링크 추가, 임시 사용자 데이터 설정

One-HyeWon commented on src/app/(route)/(afterLogin)/_component/Sidebar.tsx:
```diff
@@ -0,0 +1,76 @@
+    <aside className="fixed w-64 h-screen border-r border-gray-700">
+      <div className="flex flex-col h-full p-4">
+        <div className="p-4">
+          <svg viewBox="0 0 24 24" className="h-8 w-8 text-white" fill="currentColor">
+            <g>
+              <path d="M18.244 2.25h3.308l-7.227 8.26 8.502 11.24H16.17l-5.214-6.817L4.99 21.75H1.68l7.73-8.835L1.254 2.25H8.08l4.713 6.231zm-1.161 17.52h1.833L7.084 4.126H5.117z"></path>
+            </g>
+          </svg>
```
만약 고정되는 이미지 / 로고라면, image 태그를 따로 사용 하시는것을 추천드립니다
Next는 최적화 된 내장 컴포넌트가 있어서 확인 해보시면 좋을 것 같아요!
그리고 image path들은 상수 파이로 관리 하시는 것을 추천 드립니다
constant 파일에 imagePath.ts를 만들어 이미지 / 아이콘 경로들을 추가해두고 불러와서 사용하는 것도 좋을 것 같아요!

One-HyeWon commented on src/app/(route)/(afterLogin)/_component/Sidebar.tsx:
```diff
@@ -0,0 +1,76 @@
+            </g>
+          </svg>
+        </div>
+        <nav className="flex-1">
+          <ul className="space-y-4">
+            <li>
+              <Link href="/home" className="flex items-center gap-4 text-xl hover:bg-gray-800 px-4 py-3 rounded-full">
+                <AiFillHome className="w-7 h-7" />
+                <span>홈</span>
+              </Link>
+            </li>
+            <li>
+              <Link href="/messages" className="flex items-center gap-4 text-xl hover:bg-gray-800 px-4 py-3 rounded-full">
+                <RiMailLine className="w-7 h-7" />
+                <span>쪽지</span>
+              </Link>
+            </li>
+            <li>
+              <Link href="/profile" className="flex items-center gap-4 text-xl hover:bg-gray-800 px-4 py-3 rounded-full">
+                <CgProfile className="w-7 h-7" />
+                <span>프로필</span>
+              </Link>
+            </li>
+          </ul>
+        </nav>
```
list 태그로 구현 하신 이유가 있으신가요?

One-HyeWon commented on src/app/(route)/(afterLogin)/home/PostModal.tsx:
```diff
@@ -0,0 +1,69 @@
+const PostModal: React.FC<PostModalProps> = ({ onSubmit }) => {
+  const [postContent, setPostContent] = useState('');
+  const [isClient, setIsClient] = useState(false);
+  const router = useRouter(); // 클라이언트에서만 사용할 수 있음
```
주석을 달아주신 것 처럼 use client는 client 단에서 모든 요소들이 렌더링 됩니다! 따라서 Next.js를 사용하는 가장 큰 이유인 SSR을 최대한 활용 하시려면, 'use client'는 최대한 컴포넌트 하나하나로 분리하여 사용해주시는 것이 좋아요!

지금처럼 하나의 파일에 렌더링 되어야 할 구성 요소들이 많다면, router.back을 사용하는 컴포넌트들은 따로 분류하여 만들어주신 후, 이 파일에서 Import 하여 사용 하시면 이 파일은 SSR이 가능해지니 다른 text들은 서버단에서 미리 렌더링 될 수 있습니다!

One-HyeWon commented on src/app/(route)/(afterLogin)/home/PostModal.tsx:
```diff
@@ -0,0 +1,69 @@
+    }
+  };
+
+  if (!isClient) {
+    return null; // 클라이언트에서만 렌더링
+  }
```
해당 구문을 추가해주신 이유가 있으실까요?

One-HyeWon commented on src/app/(route)/(afterLogin)/home/layout.tsx:
```diff
@@ -1,102 +1,85 @@
+            <div className="flex justify-between items-center mb-4">
+              <h2 className="text-xl font-bold">게시글 작성</h2>
+              <button onClick={handleCloseModal} className="text-gray-500 hover:text-gray-700">
+                <svg className="w-6 h-6" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
+                  <path strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="M6 18L18 6M6 6l12 12"></path>
+                </svg>
```
위와 동일합니다

One-HyeWon commented on src/app/(route)/(afterLogin)/home/layout.tsx:
```diff
@@ -1,102 +1,85 @@
+        {children}
+      </main>
+      {isModalOpen && (
+        <div className="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center">
+          <div className="bg-white text-black p-6 rounded-lg w-96">
+            <div className="flex justify-between items-center mb-4">
```
컨테이너 역할을 하는 div들이 많으면, container 컴포넌트를 따로 만들어주시는 것도 좋을 것 같아요!
그러면 어느 div 태그가 어느 요소들을 담고 있는지 더 쉽게 알 수 있을 것 같습니다

One-HyeWon commented on src/app/(route)/(afterLogin)/home/page.tsx:
```diff
@@ -1,114 +1,331 @@
+    const post: Post = {
+      id: posts.length + 1,
+      author: "Myself",
+      authorImage: "/profile-placeholder.png",
```
미리 아무 이미지나 가지고 오셔서 확인 해보셔도 좋을 것 같아요

heyn2 committed:
게시물 작성 모달

heyn2 committed:
Xcellent-FE\src\app\(route)\(afterLogin)\_component\PostModal.tsx로 이동

heyn2 committed:
게시물 상태 관리 및 좋아요, 리트윗, 댓글 기능 추가

- 게시물 목록을 상태로 관리하고 새로운 게시물이 추가될 때 업데이트되도록 수정
- 각 게시물에 대해 좋아요 토글, 리트윗 증가, 댓글 작성 기능 구현
- HomePage 컴포넌트에 필요한 이벤트 핸들러(onLike, onRetweet, onCommentSubmit) 전달

heyn2 committed:
게시글 작성 모달에서 작성된 게시물이 화면에 표시되도록 수정

- 이제 PostModal에서 작성된 게시물이 실제로 화면에 표시되도록 posts 배열에 추가
- 각 게시물에 대해 좋아요 토글, 리트윗 증가, 댓글 작성 기능 구현
- HomePage 컴포넌트에 필요한 이벤트 핸들러(onLike, onRetweet, onCommentSubmit) 전달

heyn2 committed:
Feat: #4 / 게시물 컴포넌트 생성

heyn2 committed:
충돌 해결

heyn2 committed:
작업 중인 변경 사항 커밋

heyn2 committed:
홈페이지 ui 수정

heyn2 committed:
develop-후-충돌

heyn2 committed:
충돌 해결
